File: Java/src/TransposeMatrix.java
/**
 * Given a matrix A, return the transpose of A.
 * <p>
 * The transpose of a matrix is the matrix flipped over it's main diagonal, switching the row and column indices of the matrix.
 * <p>
 * Example 1:
 * <p>
 * Input: [[1,2,3],[4,5,6],[7,8,9]]
 * Output: [[1,4,7],[2,5,8],[3,6,9]]
 * Example 2:
 * <p>
 * Input: [[1,2,3],[4,5,6]]
 * Output: [[1,4],[2,5],[3,6]]
 * <p>
 * Note:
 * <p>
 * 1 <= A.length <= 1000
 * 1 <= A[0].length <= 1000
 * <p>
 * Result: Accepted.
 */
public class TransposeMatrix {

    public int[][] transpose(int[][] A) {
        int[][] tmp = new int[A[0].length][A.length];
        for (int i = 0; i < A[0].length; i++) {
            for (int j = 0; j < A.length; j++) {
                tmp[i][j] = A[j][i];
            }
        }

        return tmp;
    }

}

File: Kotlin/src/TransposeMatrix.kt
/**
 * Given a matrix A, return the transpose of A.
 *
 * The transpose of a matrix is the matrix flipped over it's main diagonal, switching the row and column indices of the matrix.
 *
 * Example 1:
 *
 * Input: [[1,2,3],[4,5,6],[7,8,9]]
 * Output: [[1,4,7],[2,5,8],[3,6,9]]
 * Example 2:
 *
 * Input: [[1,2,3],[4,5,6]]
 * Output: [[1,4],[2,5],[3,6]]
 *
 * Note:
 *
 * 1 <= A.length <= 1000
 * 1 <= A[0].length <= 1000
 *
 * Result: Accepted.
 */
class TransposeMatrix {

    fun transpose(A: Array<IntArray>): Array<IntArray> {
        val tmp = Array(A[0].size) { IntArray(A.size) }
        for (i in 0 until A[0].size) {
            for (j in A.indices) {
                tmp[i][j] = A[j][i]
            }
        }

        return tmp
    }

}
